packages/mongo/changestream_observe_driver.js

```diff
@@ -0,0 +1,623 @@
+          await write.committed();
+        }
+      });
+    }
+  }
+
+  _handleInsert(id, doc) {
+    // Apply projection and check if document matches our criteria
+    const matches = this._matcher ? this._matcher.documentMatches(doc).result : true;
+    if (matches) {
+      const projectedDoc = this._projectionFn ? this._projectionFn(doc) : doc;
+      this._sendMultiplexerAdded(id, projectedDoc);
+    }
+  }
+
+  _handleUpdate(id, newDoc, oldDoc) {
+    // Determine which state (before/after) matches the cursor selector
+    const matchesAfter = this._matcher
+      ? this._matcher.documentMatches(newDoc || {}).result
+      : true;
+
+    // If MongoDB delivers the pre-image we can rely on it. Otherwise fall back to
+    // the multiplexer cache to infer whether we were previously tracking the doc.
+    const cachedDoc = this._multiplexer?._cache?.docs?.get?.(id);
+    const matchesBefore = oldDoc
+      ? (this._matcher ? this._matcher.documentMatches(oldDoc).result : true)
+      : !!cachedDoc;
+
+    if (matchesAfter) {
+      if (!matchesBefore) {
+        // Document wasn't previously in the result set and now matches – emit added
+        const projectedDoc = this._projectionFn ? this._projectionFn(newDoc) : newDoc;
+        this._sendMultiplexerAdded(id, projectedDoc);
+        return;
+      }
+
+      if (newDoc) {
+        // Compute the changed fields using the available pre-image or the cached doc
+        const oldDocForDiff = oldDoc || (cachedDoc ? { ...cachedDoc } : null);
+        if (oldDocForDiff) {
+          const projectedNew = this._projectionFn ? this._projectionFn(newDoc) : newDoc;
+          const projectedOld = this._projectionFn ? this._projectionFn(oldDocForDiff) : oldDocForDiff;
+          const changedFields = DiffSequence.makeChangedFields(projectedNew, projectedOld);
+
+          if (Object.keys(changedFields).length > 0) {
+            const transformedDoc = replaceTypes(changedFields, replaceMongoAtomWithMeteor);
+            this._multiplexer.changed(id, transformedDoc);
+          }
+          return;
+        }
+
+        // Without a pre-image we can't diff reliably; fall back to sending full doc
+        const projectedDoc = this._projectionFn ? this._projectionFn(newDoc) : newDoc;
+        const transformedDoc = replaceTypes(projectedDoc, replaceMongoAtomWithMeteor);
+        this._multiplexer.changed(id, transformedDoc);
+      }
+      return;
+    }
+
+    if (matchesBefore) {
+      // Document left the result set
+      this._multiplexer.removed(id);
+    }
+    // Otherwise the document didn't match before or after, so no-op
+  }
+
+  _handleReplace(id, newDoc, oldDoc) {
+    // Handle replace similar to update
+    this._handleUpdate(id, newDoc, oldDoc);
+  }
+
+  _handleDelete(id) {
+    const docs = this._multiplexer?._cache?.docs;
+    const hasDoc = typeof docs?.has === 'function'
+      ? docs.has(id)
+      : typeof docs?.get === 'function' && docs.get(id) !== undefined;
```

Comment:
Typing here would help, but we have too many `?.`s here I think (e.g., the first one), and the duck typing of the methods seems unexpected - why is `has` not always present?